src/main/java/com/redhat/devtools/lsp4ij/features/diagnostics/LSPDiagnosticAnnotator.java

```diff
@@ -66,28 +91,86 @@ public void doApply(@NotNull PsiFile file, Boolean applyAnnotator, @NotNull Anno
+                        var ds = data.getDiagnosticsForServer();
+                        // Loop for LSP diagnostics to transform it to Intellij annotation.
+                        // TODO: When running for bulk inspection, we need to wait for diagnostics here; not sure how to
+                        //  do that since it seems it's all based on publishDiagnostics notifications
+                        for (Diagnostic diagnostic : ds.getDiagnostics()) {
+                            ProgressManager.checkCanceled();
+                            createAnnotation(diagnostic, document, file, ds, holder);
+                        }
+                    }
+                }
+            } finally {
+                if (needsDisconnect) {
+                    ls.disconnect(fileUri, false);
+                }
+            }
+        }
+    }
+
+    private boolean connectIfNecessary(@NotNull LanguageServerWrapper ls, @NotNull PsiFile file) {
+        // Not while indexing
+        if (ProjectIndexingManager.canExecuteLSPFeature(file) != ExecuteLSPFeatureStatus.NOW) {
+            return false;
+        }
+
+        URI fileUri = LSPIJUtils.toUri(file);
+        if (!ls.isConnectedTo(fileUri)) {
+            VirtualFile virtualFile = file.getVirtualFile();
+            Document document = virtualFile != null ? LSPIJUtils.getDocument(virtualFile) : null;
+            if (document != null) {
+                CompletableFuture<LanguageServer> connectionResultFuture = ls.connect(virtualFile, document);
+
+                try {
+                    // Wait until the future is finished and stop the wait if there are some ProcessCanceledException.
+                    waitUntilDone(connectionResultFuture, file);
+                } catch (ProcessCanceledException ignore) {
+                    //Since 2024.2 ProcessCanceledException extends CancellationException so we can't use multicatch to keep backward compatibility
+                    //TODO delete block when minimum required version is 2024.2
+                    return false;
+                } catch (CancellationException ignore) {
+                    return false;
+                } catch (ExecutionException e) {
+                    LOGGER.error("Error while connecting file '{}' to language server '{}'.", file, ls, e);
+                    return false;
+                }
+
+                ProgressManager.checkCanceled();
+                if (isDoneNormally(connectionResultFuture)) {
+                    LanguageServer languageServer = connectionResultFuture.getNow(null);
+                    return languageServer != null;
                 }
             }
         }
+
+        return false;
     }
 
     private static void createAnnotation(@NotNull Diagnostic diagnostic,
                                          @NotNull Document document,
-                                         @Nullable PsiFile file,
+                                         @NotNull PsiFile file,

```

Comment:
All existing callers of this had already guaranteed a non-null file, so I changed the parameter nullability annotation accordingly since the code is using it as if it were non-null anyway.